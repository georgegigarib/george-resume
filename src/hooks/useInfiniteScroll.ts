import { useEffect, useRef } from 'react'

const useInfiniteScroll = (containerId: string) => {
  const containerRef = useRef<HTMLDivElement>(null)

  useEffect(() => {
    const container = containerRef.current
    if (!container) return

    let animationFrame: number
    let scrollPosition = 0

    try {
      // Duplicamos el contenido para permitir un scroll infinito
      const originalContent = container.innerHTML
      container.innerHTML += originalContent

      // Calculamos el ancho total para ajustar el desplazamiento
      const scrollWidth = container.scrollWidth / 2

      const scrollStep = () => {
        scrollPosition += 0.5 // Ajusta la velocidad del desplazamiento aquÃ­
        if (scrollPosition >= scrollWidth) {
          scrollPosition = 0
        }
        container.scrollLeft = scrollPosition
        animationFrame = requestAnimationFrame(scrollStep)
      }

      scrollStep()
    } catch (error) {
      console.error('Error in useInfiniteScroll:', error)
    }

    // Limpieza al desmontar
    return () => {
      if (animationFrame) cancelAnimationFrame(animationFrame)
    }
  }, [containerId])

  return containerRef
}

export default useInfiniteScroll
