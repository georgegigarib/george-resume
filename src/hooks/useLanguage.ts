import { useTranslation } from 'react-i18next';

export const useLanguage = () => {
  const { i18n } = useTranslation();

  const changeLanguage = (lng: string) => {
    i18n.changeLanguage(lng);
    localStorage.setItem('language', lng);
  };

  const loadLanguageFromStorage = () => {
    const savedLanguage = localStorage.getItem('language') || 'en';
    i18n.changeLanguage(savedLanguage);
  };

  return { changeLanguage, loadLanguageFromStorage };
};
