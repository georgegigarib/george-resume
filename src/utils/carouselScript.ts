export default function initCarousel(
  speed: number = -200,
  width: string = '100px',
  height: string = '180px',
  rotate: boolean = true,
  circleRadius: number = 2
): void {
  const odrag = document.getElementById('drag-container') as HTMLElement
  const ospin = document.getElementById('spin-container') as HTMLElement

  // Verificación para asegurar que los elementos existan en el DOM
  if (!odrag || !ospin) {
    console.error('Elements not found')
    return
  }

  const aImg: HTMLCollectionOf<HTMLImageElement> = ospin.getElementsByTagName('img')
  const aVid: HTMLCollectionOf<HTMLVideoElement> = ospin.getElementsByTagName('video')
  const aEle: (HTMLImageElement | HTMLVideoElement)[] = [...aImg, ...aVid]

  // Configuración global
  const rotateSpeed = speed
  const imgWidth = width
  const imgHeight = height
  const autoRotate = rotate
  const radius = circleRadius

  // Tamaño de imágenes y contenedor
  ospin.style.width = `${imgWidth}`
  ospin.style.height = `${imgHeight}`

  const ground = document.getElementById('ground') as HTMLElement
  if (ground) {
    ground.style.width = `${radius * 20}px`
    ground.style.height = `${radius * 20}px`
  }

  // Inicializar elementos en círculo
  for (let i = 0; i < aEle.length; i++) {
    aEle[i].style.transform = `rotateY(${i * (360 / aEle.length)}deg) translateZ(${radius}vh)`
    aEle[i].style.transition = 'transform 1s'
    aEle[i].style.transitionDelay = `${(aEle.length - i) / 4}s`
  }

  let tX = 0
  let tY = 10
  let desX = 0
  let desY = 0

  const timers = new Map<HTMLElement, ReturnType<typeof setTimeout>>()

  function applyTransform(): void {
    if (tY > 180) tY = 180
    if (tY < 0) tY = 0
    odrag.style.transform = `rotateX(${-tY}deg) rotateY(${tX}deg)`
  }

  // Auto-rotación
  if (autoRotate) {
    ospin.style.animation = `${rotateSpeed > 0 ? 'spin' : 'spinRevert'} ${Math.abs(rotateSpeed)}s infinite linear`
  }

  // Eventos de interacción con el mouse
  document.onpointerdown = function (e: PointerEvent): false {
    const currentTimer = timers.get(odrag)
    if (currentTimer) clearInterval(currentTimer)

    const sX = e.clientX
    const sY = e.clientY

    document.onpointermove = function (e: PointerEvent) {
      const nX = e.clientX
      const nY = e.clientY
      desX = nX - sX
      desY = nY - sY
      tX += desX * 0.03
      tY += desY * 0.01
      applyTransform()
    }

    document.onpointerup = function () {
      const timer = setInterval(() => {
        desX *= 0.95
        desY *= 0.95
        tX += desX * 0.03
        tY += desY * 0.01
        applyTransform()
        if (Math.abs(desX) < 0.5 && Math.abs(desY) < 0.5) {
          const currentTimer = timers.get(odrag)
          if (currentTimer) clearInterval(currentTimer)
        }
      }, 13)

      // Guardamos el temporizador en el Map
      timers.set(odrag, timer)

      document.onpointermove = document.onpointerup = null
    }

    return false
  }
}
