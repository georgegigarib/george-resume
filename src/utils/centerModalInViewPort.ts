export const centerModalInViewport = (isMobile?: boolean) => {
  const viewportWidth = window.innerWidth
  const viewportHeight = window.innerHeight
  const modalWidth = isMobile ? viewportWidth - 80 : Math.min(450, viewportWidth * 0.9)
  const modalHeight = 400

  const left = (viewportWidth - modalWidth) / 2
  const top = (viewportHeight - modalHeight) / 2

  return { top, left }
}
