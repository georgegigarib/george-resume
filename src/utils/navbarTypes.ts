export interface boxType {
    startRow: number,
    spanRow: number,
    startColumn: number,
    spanColumn: number,
    content: string
}