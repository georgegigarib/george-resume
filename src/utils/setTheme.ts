export function setTheme(isDark: boolean): void {
  if (isDark) {
    document.documentElement.classList.add('dark')
  } else {
    document.documentElement.classList.remove('dark')
  }
}

export function isDarkTheme(): boolean {
  return document.documentElement.classList.contains('dark')
}
