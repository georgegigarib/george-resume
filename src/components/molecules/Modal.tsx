import React, { useState, useEffect } from 'react'
import { motion, AnimatePresence } from 'framer-motion'
import { IconButton } from '@mui/material'
import CloseIcon from '@mui/icons-material/Close'
import { useIsMobile } from '@/hooks/useIsMobile'
import { centerModalInViewport } from '@/utils/centerModalInViewPort'
import clsx from 'clsx'

interface ModalProps {
  trigger: React.ReactNode
  children: React.ReactNode
  layoutId: string
  normalWidth?: string
  mobileHeight?: string
  normalHeight?: string
  className?: string
}

const Modal: React.FC<ModalProps> = ({
  layoutId,
  trigger,
  children,
  normalWidth = '450px',
  mobileHeight = '350px',
  normalHeight = '500px',
  className
}) => {
  const [isOpen, setIsOpen] = useState(false)
  const [modalPosition, setModalPosition] = useState({ top: 0, left: 0 })
  const isMobile = useIsMobile()

  useEffect(() => {
    if (isOpen) {
      setModalPosition(centerModalInViewport(isMobile))
    }
    const handleResize = () => setModalPosition(centerModalInViewport(isMobile))
    window.addEventListener('resize', handleResize)
    return () => window.removeEventListener('resize', handleResize)
  }, [isOpen, isMobile])

  const handleClose = () => setIsOpen(false)

  const handleOutsideClick = (e: React.MouseEvent<HTMLDivElement>) => {
    if (e.target === e.currentTarget) {
      handleClose()
    }
  }

  return (
    <>
      <motion.div
        className="h-full w-full cursor-pointer"
        layoutId={layoutId}
        onClick={() => setIsOpen(true)}
      >
        {trigger}
      </motion.div>

      <AnimatePresence mode="popLayout">
        {isOpen && (
          <>
            <motion.div
              transition={{ ease: 'circIn', duration: 0.5 }}
              className="fixed inset-0 bg-black bg-opacity-50 backdrop-blur-sm flex justify-center items-center z-[20]"
              onClick={handleOutsideClick}
              initial={{ opacity: 0 }}
              animate={{ opacity: 1 }}
              exit={{ opacity: 0 }}
            >
              <motion.div
                layoutId={layoutId}
                className={clsx(
                  'relative rounded-xl p-4',
                  `top-[${modalPosition.top}] left-[${modalPosition.left}]`,
                  className
                )}
                style={{
                  width: isMobile ? 'calc(100% - 80px)' : normalWidth,
                  height: isMobile ? mobileHeight : normalHeight,
                  position: 'fixed'
                }}
                initial={{ opacity: 0 }}
                animate={{ opacity: 1 }}
                exit={{ opacity: 0 }}
              >
                <div className="absolute top-4 right-4">
                  <IconButton onClick={handleClose} aria-label="close">
                    <CloseIcon
                      fontSize="medium"
                      style={{ fill: 'white' }}
                      className="p-1 hover:bg-gray-700 rounded-full"
                    />
                  </IconButton>
                </div>
                {children}
              </motion.div>
            </motion.div>
          </>
        )}
      </AnimatePresence>
    </>
  )
}

export default Modal
