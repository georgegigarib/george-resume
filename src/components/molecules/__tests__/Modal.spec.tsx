import { render, fireEvent, screen, waitFor } from '@testing-library/react'
import { useIsMobile } from '@/hooks/useIsMobile'
import '@testing-library/jest-dom'
import { vi, describe, Mock, it, beforeEach, expect } from 'vitest'
import { centerModalInViewport } from '@/utils/centerModalInViewPort'
import Modal from '../Modal'

vi.mock('@/utils/centerModalInViewport')
vi.mock('@/hooks/useIsMobile')

describe('Modal Component', () => {
  beforeEach(() => {
    vi.clearAllMocks()
    ;(centerModalInViewport as Mock).mockReturnValue({ top: 100, left: 100 })
  })

  it('matches snapshot', () => {
    const { asFragment } = render(
      <Modal trigger={<button>Open</button>} layoutId="some-layoutid">
        <div>Modal Content</div>
      </Modal>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('opens modal when trigger is clicked', () => {
    render(
      <Modal trigger={<button>Open</button>} layoutId="some-layoutid">
        <div>Modal Content</div>
      </Modal>
    )

    fireEvent.click(screen.getByText('Open'))
    expect(screen.getByText('Modal Content')).toBeInTheDocument()
  })

  it('closes modal when close button is clicked', async () => {
    render(
      <Modal trigger={<button>Open</button>} layoutId="some-layoutid">
        <div>Modal Content</div>
      </Modal>
    )

    fireEvent.click(screen.getByText('Open'))
    expect(screen.getByText('Modal Content')).toBeInTheDocument()

    fireEvent.click(screen.getByTestId('CloseIcon'))

    await waitFor(() => {
      expect(screen.queryByText('Modal Content')).not.toBeInTheDocument()
    })
  })

  it('calls centerModalInViewport to center modal when opened', () => {
    const mockPosition = { top: 100, left: 100 }
    ;(centerModalInViewport as Mock).mockReturnValue(mockPosition)
    ;(useIsMobile as Mock).mockReturnValue(false)

    render(
      <Modal trigger={<button>Open</button>} layoutId="some-layoutid">
        <div>Modal Content</div>
      </Modal>
    )

    fireEvent.click(screen.getByText('Open'))

    expect(centerModalInViewport).toHaveBeenCalledWith(false)
    expect(screen.getByText('Modal Content')).toBeInTheDocument()
  })

  it('adjusts modal position on resize', () => {
    const mockPosition = { top: 100, left: 100 }
    ;(centerModalInViewport as Mock).mockReturnValue(mockPosition)

    render(
      <Modal trigger={<button>Open</button>} layoutId="some-layoutid">
        <div>Modal Content</div>
      </Modal>
    )

    fireEvent.click(screen.getByText('Open'))

    // Simulate window resize
    fireEvent.resize(window)
    expect(centerModalInViewport).toHaveBeenCalledTimes(2)
  })

  it('adapts to mobile layout when isMobile is true', () => {
    ;(useIsMobile as Mock).mockReturnValue(true)

    render(
      <Modal trigger={<button>Open</button>} layoutId="some-layoutid">
        <div>Modal Content</div>
      </Modal>
    )

    fireEvent.click(screen.getByText('Open'))
    expect(centerModalInViewport).toHaveBeenCalledWith(true)
  })
})
