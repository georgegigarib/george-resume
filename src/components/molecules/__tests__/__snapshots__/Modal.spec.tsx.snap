// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-full w-full cursor-pointer"
  >
    <button>
      Open
    </button>
  </div>
</DocumentFragment>
`;
